packages/ai-hugging-face/src/node/huggingface-language-model.ts

```diff
@@ -0,0 +1,126 @@
+// *****************************************************************************
+// Copyright (C) 2024 EclipseSource GmbH.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import {
+    LanguageModel,
+    LanguageModelRequest,
+    LanguageModelRequestMessage,
+    LanguageModelResponse,
+    LanguageModelStreamResponsePart,
+    LanguageModelTextResponse,
+    MessageActor
+} from '@theia/ai-core';
+import { CancellationToken } from '@theia/core';
+import { HfInference } from '@huggingface/inference';
+
+export const HuggingFaceModelIdentifier = Symbol('HuggingFaceModelIdentifier');
+
+function toHuggingFacePrompt(messages: LanguageModelRequestMessage[]): string {
+    if (messages.length === 1) {
+        return messages[0].query;
+    }
+    return messages.map(message => `${toRoleLabel(message.actor)}: ${message.query}`).join('\n');
+}
+
+function toRoleLabel(actor: MessageActor): string {
+    switch (actor) {
+        case 'user':
+            return 'User';
+        case 'ai':
+            return 'Assistant';
+        case 'system':
+            return 'System';
+        default:
+            return '';
+    }
+}
+
+export class HuggingFaceModel implements LanguageModel {
+    private hfInference: HfInference;
+
+    /**
+     * @param id the unique id for this language model. It will be used to identify the model in the UI.
+     * @param model the model id as it is used by the Hugging Face API
+     * @param apiKey function to retrieve the API key for Hugging Face
+     */
+    constructor(public readonly id: string, public model: string, public apiKey: () => string | undefined) {
+        const token = this.apiKey();
+        if (!token) {
+            throw new Error('Please provide a Hugging Face API token.');
+        }
```

Comment:
We should not throw errors for expected code paths. This is even the default code path as the api key will be `undefined` by default. Therefore this will show up as an error for all users in their logs.

If we don't want to be able to configure hugging face models without keys, then the `huggingface-language-model-manager` should not even create them.

Note that if the user deletes the `apiKey` in their preferences, you will still use it as the `hfInference` is only created once here. In fact the whole dynamic `apiKey` provider is not utilized in this implementation as it's only called once.

I would like to suggest to either make the `apiKey` static here and making sure that the models are removed/recreated if the `apiKey` changes or refactor them to being able to handle non existent api keys at request time, like in the open ai language model implementation.